### PR TITLE
Client spec

### PR DESCRIPTION
**Pull request by morrigan**

### CODE REVIEW

File: spec/components/Web3API_Client.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+```
+interface Web3APIClient {
+  redirects(): UriRedirect[];
+  query(options: QueryApiOptions): QueryApiResult;
+}
+
+type UriRedirect {
+ // Redirect from this URI or set of URIs specified by the regex
+ from: Uri | RegExp;
```
Comment (mpetrunic) [inline spec/components/Web3API_Client.md:19]: I guess string?

File: spec/components/Web3API_Client.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+type QueryApiResult {
+ /**
+   * Query result data. The type of this value is a named map,
+   * where the key is the method's name, and value is the [[InvokeApiResult]]'s data.
+   * This is done to allow for parallel invocations within a
+   * single query document. In case of method name collisions,
+   * a postfix of `_0` will be applied, where 0 will be incremented for
+   * each occurrence. If undefined, it means something went wrong.
+ */
+  data?: [[InvokeApiResult]];
```
Comment (mpetrunic) [inline spec/components/Web3API_Client.md:42]: ain't this suppose to be map or tuple?
Not entirely sure how to hanle "any" value in other languages. I wonder if there are some restrictions imposed by msg pack on type of result.

File: spec/components/Web3API_Client.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+Web3API client is intended for application developers that can use the client as a single library in their (d)apps to interact with various decentralized services using Web3API's packages.
+
+Currently, Web3API client is available in the following languages:
+ - JavaScript (TypeScript)
+
+## Client interface
+The following snippet contains the interface that client should implement.
+
+```
```
Comment (mpetrunic) [inline spec/components/Web3API_Client.md:11]: Maybe emphasize that those are written in typescript. Some more generic IDL language might be better.

File: spec/components/Web3API_Client.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+### Initialization
+Client is initialized using `Web3APIClient` class that should support receiving:
+
+#### `redirects` array
+This in an unbounded array containing `UriRedirect` objects.
+Redirects are used for resolving Web3API URIs to decentralized protocols and their nodes.
+ 
+Upon initizalition it is filled by default redirects:
```
Comment (mpetrunic) [inline spec/components/Web3API_Client.md]:
```suggestion
Upon initizalition it is filled with default redirects:
```

File: spec/components/Web3API_Client.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+ - converting the query string into a query document
+ - parsing the query by calling parse-query algorithm
+ - processing API invocations on loaded Web3APIs
+ - returning from API result `data` and `errors`
+
+#### `redirects` function
+
+Returns URI redirects array set upon initialization. 
+
+## URI resolution algorithm
```
Comment (mpetrunic) [inline spec/components/Web3API_Client.md:93]: separate file?

File: spec/components/Web3API_Client.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+Client is initialized using `Web3APIClient` class that should support receiving:
+
+#### `redirects` array
+This in an unbounded array containing `UriRedirect` objects.
+Redirects are used for resolving Web3API URIs to decentralized protocols and their nodes.
+ 
+Upon initizalition it is filled by default redirects:
+ - IPFS - required for downloading Web3API packages
+ - Ethereum - required for smart contract interactions
+ - ENS - required for resolving domain to IPFS hashes
```
Comment (mpetrunic) [inline spec/components/Web3API_Client.md]: Package interface and specification for each of preloaded redirects. If I'm not mistaken those must be implemented in every client language? Or if they depend on host methods, than host methods must be specified.

File: spec/components/Web3API_Client.md (hunk 1 above)
Comment (mpetrunic) [reply spec/components/Web3API_Client.md:19]: Oh I see Uri interface down below. You should put it here before first usage or link it.

File: spec/components/Web3API_Client.md (hunk 5 above)
Comment (morrigan) [reply spec/components/Web3API_Client.md:93]: Yeah, not sure. Maybe all algorithms should be separated if all are going to be covered in the same way. Leaving it open

File: spec/components/Web3API_Client.md (hunk 6 above)
Comment (morrigan) [reply spec/components/Web3API_Client.md]: I left them out intentionally but good question. Should the default redirects be the same and implemented in every client @dOrgJelli making them part of the spec?

File: spec/components/Web3API_Client.md (hunk 1 above)
Comment (morrigan) [reply spec/components/Web3API_Client.md:19]: I don't think markdown has so many options for linking xD Moving.

**Pull request closed without merging**